Cubical Complexes

GeorgeR227 opened the pull request:
This PR introduces support for 2D cubical complexes. Namely, these are rectilinear grids that can be defined implicitly. These types of grids are useful for simulations where horizontal and vertical directions are distinguished in some way (e.g. gravity is only vertical).

The implicit mesh, at the time of this comment, only depends on an explicit list of vertex points. All other elements, including dual elements, are defined from these primal points. Vertices and quadrilaterals are given both x,y coordinate labels and sequential coordinate labels, which are given from left to right and top to bottom. 

Desired features would be:
- [ ] Implicit treatment of point coordinates (explicit lists of points are replaced by functions providing coordinates)
- [ ] KernelAbstractions support for all DEC operators (to support multi-threading/GPU)
- [ ] Proper treatment of periodic boundary conditions (meaning automatically generated and easy to apply)
- [ ] Support for distributed computing (eased since the mesh is a grid and all inform is implicit and can be easy generated)

GeorgeR227 committed:
Began work on cubical complexes

GeorgeR227 committed:
Added quad to vertex map

GeorgeR227 committed:
Added exterior derivatives

Still needs better handling of orientation for quads and dual meshes

GeorgeR227 committed:
Added plotting code and relative orientations

This adds support for mesh plotting through Makie and proper computation of exterior derivatives. Next steps will be dual mesh generation.

GeorgeR227 committed:
Switched to implicit grid

Now meshes only store point data and derive all other information implicitly.

GeorgeR227 committed:
Added several DEC operators

Dual wedge 01 needs to be added and further tests on operators need to be encoded. Also, built in support for periodic boundaries is needed.

GeorgeR227 committed:
Created diffusion and advection tests

Also fixed Hodge star 1 and added more tests.

GeorgeR227 committed:
Added more sharps/flat + periodic work

GeorgeR227 committed:
Fixed periodic advection

GeorgeR227 committed:
Taylor vortices in cubical and implicit periodic conditions

GeorgeR227 committed:
Added in more general operators

GeorgeR227 committed:
Introduce forms as multidimensional tensors

GeorgeR227 committed:
Added more tensorized kernel code

GeorgeR227 committed:
Better organization and kernel operators in physics

GeorgeR227 committed:
Switching mesh code to use CartesianIndices for better kernel integration

GeorgeR227 committed:
Make further changes to support coordinates

GeorgeR227 committed:
Replaced Cartesian indices with tuples

GeorgeR227 committed:
Worked on kernels, still need interpolation

GeorgeR227 committed:
Added final operators needed + tests

GeorgeR227 committed:
Add periodic condition kernel and add benchmarks

lukem12345 commented on docs/src/euler.md:
```diff
@@ -70,9 +70,9 @@ function plot_dvf(sd, X; ls=1f0, title="Dual Vector Field")
 end
 
-sharp_pp = ♯_mat(sd, AltPPSharp())
+sharp_pd = ♯_mat(sd, AltPPSharp())
```
What's going on here? It looks like a primal primal sharp is being assigned a primal dual name.

GeorgeR227 committed:
Improved mesh element iterators and caught up matrix operators

GeorgeR227 committed:
Update incompressible with matrix/kernel code

GeorgeR227 committed:
Fixed hodge 2, added all periodic and added Lie advection test

GeorgeR227 committed:
Experiments with high-res advection

GeorgeR227 committed:
Basic upwinding advection in 2D

GeorgeR227 committed:
Added QUICK scheme in 2D, basic implementation

GeorgeR227 committed:
Added WENO test

GeorgeR227 committed:
Basic advection working

GeorgeR227 committed:
Working WENO in 1D

Will need to switch off Euler since it is unstable

GeorgeR227 committed:
Completed WENO simulation in 1D

This incorporates the smoothness indicators as part of the computation.

GeorgeR227 committed:
Added concrete dual mesh and work on 2D WENO

GeorgeR227 committed:
Fixed advection schemes

Also added lazy dual point list

GeorgeR227 committed:
Added more iterators for edges

GeorgeR227 committed:
Working WENO5 for 2D linear advection

GeorgeR227 committed:
Completed WENO5 with upwinding direction

GeorgeR227 committed:
Updated to WENO5-Z

GeorgeR227 committed:
Improved WENO code, WENO 9, Burger's equation, shear advection test

GeorgeR227 committed:
Added new uniform cubical code and got working cavity flow experiment

GeorgeR227 committed:
Added sim parameters for cavity flow

GeorgeR227 committed:
Got basic versions of NS working

GeorgeR227 committed:
Added attempt at compressible cavity flow

wkunzendorf-hub committed:
Turbulent Cavity Flow Changes

This change adds code, plots, and flow data into the CubicalTest folder from my current Re=10K simulation. Full data from each velocity and pressure step is not included.

wkunzendorf-hub committed:
Update Cavity Flow

Further additions to the Cavity Flow file for Turbulent Flow simulations.

wkunzendorf-hub committed:
Update Cavity Flow.jl

lukem12345 commented:
@GeorgeR227 For better code management, we can create a new branch off of main, and `git cherry-pick` the appropriate commits for cubical mesh data structures and their operators. The particular physics benchmarks could be hosted in a separate repo as necessary. Thoughts?

GeorgeR227 commented:
Organization-wise that's probably the way to go. Main reason this hasn't been done yet is cause this code is being iterated along with the physics benchmarks. Essentially, cubical code is written -> physics is written -> realize bug or need for new feature -> new cubical code is written -> ad infinitum.

So for development efficiency, the code has been stuck together with the physics. So I think once we're satisfied with the 2D benchmarks for now, I'd be happy moving the physical benchmarks elsewhere.

GeorgeR227 committed:
Get new simulations working, cleaner sim files

GeorgeR227 committed:
Added tracer equation to testbed advection schemes for scalar transport

GeorgeR227 committed:
Overhaul LMNSH to use DiffEq and make Tracer bench suite

GeorgeR227 committed:
Lifted code to higher level for abstraction

GeorgeR227 committed:
Migrating general physical behavior to high level

GeorgeR227 committed:
Fixed all uniform dec tests

GeorgeR227 committed:
Added cached, efficient kernels for DEC operators

GeorgeR227 committed:
Added optimized kernels for DEC operators

GeorgeR227 committed:
Added tests for smoothing

GeorgeR227 committed:
Starting considering boundary condition-aware kernels

GeorgeR227 committed:
Added working RTI simulation

GeorgeR227 committed:
Added basic support for toml config

GeorgeR227 committed:
Kernels uses for sims and modified saving/checkpointing

GeorgeR227 committed:
Added config support for LMNSH

GeorgeR227 committed:
Added UniformCubicalComplex3D functionality

GeorgeR227 committed:
Added most of 3D cubical DEC

GeorgeR227 committed:
Changed backend retrieval

GeorgeR227 committed:
Added support for sharp_dd and flat_dp

GeorgeR227 committed:
Added basic 3D slice plotting helper

GeorgeR227 committed:
Working dual scalar heat diffusion with no-flux in 3D

GeorgeR227 committed:
Updated benchmarks file and added beginning MPI code

GeorgeR227 committed:
Basic MPI diffusion simulation running

GeorgeR227 committed:
Got working MPI Heat Diffusion simulation in 2D

GeorgeR227 committed:
Small fixes made to code

GeorgeR227 committed:
Basic worker/output topology working

Currently this only really works for scalar data. However, we expect workers to have vector data of differing dimensions. These have to be handled so that the gathering operates correctly and that the data is laid out correctly.

GeorgeR227 committed:
Progress towards worker/output with vector gathers

GeorgeR227 committed:
Significant clean up of worker/output piepline

GeorgeR227 committed:
Finished worker/output pipeline, needs testing for edges

GeorgeR227 committed:
Added signaling to IO

GeorgeR227 committed:
Updated heat with new signaling

GeorgeR227 committed:
Got 2D periodic advection working

3D heat diffusion case needs more testing, local data seems good (check exchange) but global saving/plotting is busted

GeorgeR227 committed:
Fixed bug with interior

Interior seemed to be looping over elements in the reverse way, this was leading to the visual bugs in the 3D heat diffusion problem

GeorgeR227 committed:
Got a working 3D advection simulation

Fixed final 3D heat diffusion error. Problem was too small halo causing artifacts at proccessor boundaries. Bumped up to 5 to account for 5 stages with per-stage stencil of 1.

GeorgeR227 committed:
Fixed bug in IO, edge/quad saving now works

GeorgeR227 committed:
Added worker exchange handler

Should handle packing of heterogenous data and controls halo exchange protocol

GeorgeR227 committed:
Added support for heterogenous data in exchange handler

GeorgeR227 committed:
Added ghost edges for 2D and started KH sim

Debugging of KH simulation currently in progress

GeorgeR227 committed:
Working KH simulation

MPI exchange requires further debugging and benchmarking

GeorgeR227 committed:
Added Wtime to KH

GeorgeR227 committed:
Added script driver for KH sim

GeorgeR227 committed:
Added Julia format file

GeorgeR227 committed:
Updated to allow cavity flow to work again

Commit:
Got 3D acoustic working, got AMD-aware MPI working

Commit:
Got general 2D script working and 3D WIP

Commit:
Added 3D WENO schemes

GeorgeR227 committed:
Added scripts for 3D simulations and plotting

GeorgeR227 committed:
Fixed error on line 109, updated toml for acoustic

GeorgeR227 committed:
Added wedge_dd_11, cleaned up kernel code
